ext/dom/element.c

```diff
@@ -1714,4 +1714,220 @@ PHP_METHOD(DOMElement, toggleAttribute)
+			ZVAL_NULL(OBJ_PROP_NUM(obj, 0));
+		}
+
+		if (ns != NULL && ns->href != NULL && *ns->href != '\0') {
+			ZVAL_STRING(OBJ_PROP_NUM(obj, 1), (const char *) ns->href);
+		} else {
+			ZVAL_NULL(OBJ_PROP_NUM(obj, 1));
+		}
+
+		php_dom_create_object(nodep, OBJ_PROP_NUM(obj, 2), intern);
+
+		zend_hash_next_index_insert_new(result, &zv);
+	} ZEND_HASH_FOREACH_END();
+
+	zend_hash_destroy(&prefix_to_ns_table);
+}
+
+PHP_METHOD(DOM_Element, getInScopeNamespaces)
+{
+	zval *id;
+	xmlNode *nodep;
+	dom_object *intern;
+
+	if (zend_parse_parameters_none() != SUCCESS) {
+		RETURN_THROWS();
+	}
+
+	DOM_GET_THIS_OBJ(nodep, id, xmlNodePtr, intern);
+
+	php_dom_libxml_ns_mapper *ns_mapper = php_dom_get_ns_mapper(intern);
+
+	array_init(return_value);
+	HashTable *result = Z_ARRVAL_P(return_value);
+
+	dom_element_get_in_scope_namespace_info(ns_mapper, result, nodep, intern);
+}
+
+PHP_METHOD(DOM_Element, getDescendantNamespaces)
+{
+	zval *id;
+	xmlNode *nodep;
+	dom_object *intern;
+
+	if (zend_parse_parameters_none() != SUCCESS) {
+		RETURN_THROWS();
+	}
+
+	DOM_GET_THIS_OBJ(nodep, id, xmlNodePtr, intern);
+
+	php_dom_libxml_ns_mapper *ns_mapper = php_dom_get_ns_mapper(intern);
+
+	array_init(return_value);
+	HashTable *result = Z_ARRVAL_P(return_value);
+
+	dom_element_get_in_scope_namespace_info(ns_mapper, result, nodep, intern);
+
+	xmlNodePtr cur = nodep->children;
+	while (cur != NULL) {
+		if (cur->type == XML_ELEMENT_NODE) {
+			/* TODO: this could be more optimized by updating the same HashTable repeatedly
+			 * instead of recreating it on every node. */
+			dom_element_get_in_scope_namespace_info(ns_mapper, result, cur, intern);
+
+			if (cur->children) {
+				cur = cur->children;
+				continue;
+			}
+		}
+
+		cur = php_dom_next_in_tree_order(cur, nodep);
+	}
+}
+
+PHP_METHOD(DOM_Element, rename)
+{
+	zend_string *namespace_uri, *qualified_name;
+	ZEND_PARSE_PARAMETERS_START(2, 2)
+		Z_PARAM_STR_OR_NULL(namespace_uri)
+		Z_PARAM_STR(qualified_name)
+	ZEND_PARSE_PARAMETERS_END();
+
+	zval *id;
+	dom_object *intern;
+	xmlNodePtr nodep;
+	DOM_GET_THIS_OBJ(nodep, id, xmlNodePtr, intern);
+
+	xmlChar *localname = NULL, *prefix = NULL;
+	int errorcode = dom_validate_and_extract(namespace_uri, qualified_name, &localname, &prefix);
+	if (UNEXPECTED(errorcode != 0)) {
+		php_dom_throw_error(errorcode, /* strict */ true);
+		goto cleanup;
+	}
+
+	if (nodep->type == XML_ATTRIBUTE_NODE) {
+		/* Check for duplicate attributes. */
+		xmlAttrPtr existing = xmlHasNsProp(nodep->parent, localname, namespace_uri && ZSTR_VAL(namespace_uri)[0] != '\0' ? BAD_CAST ZSTR_VAL(namespace_uri) : NULL);
+		if (existing != NULL && existing != (xmlAttrPtr) nodep) {
+			php_dom_throw_error_with_message(INVALID_MODIFICATION_ERR, "An attribute with the given name in the given namespace already exists", /* strict */ true);
+			goto cleanup;
+		}
+	} else {
+		ZEND_ASSERT(nodep->type == XML_ELEMENT_NODE);
+
+		/* Check for moving to or away from of the HTML namespace. */
```

Comment:
stomme typo... en ook een typo in de filename voor node_renaming_validation_errors ...